odoo/tools/populate.py

```diff
@@ -59,12 +61,23 @@ def get_field_variation_date(model: Model, field: Field, factor: int, series_ali
     setting duplicated records too far back in the past.
     """
     total_days = min((MAX_DATETIME - MIN_DATETIME).days, factor)
-    return SQL("""
-        %(field)s - (%(factor)s - %(series_alias)s) * floor(%(total_days)s/%(factor)s) * interval '1 days'
-    """, field=SQL.identifier(field.name),
-         factor=factor,
-         series_alias=SQL.identifier(series_alias),
-         total_days=total_days)
+    cast_type = SQL(field._column_type[1])
+    sql_code = "(%(field)s - (%(factor)s - %(series_alias)s) * floor(%(total_days)s/%(factor)s) * interval '1 days')::%(cast_type)s"
+    sql_params = {
+        'field': SQL.identifier(field.name),
+        'factor': factor,
+        'series_alias': SQL.identifier(series_alias),
+        'total_days': total_days,
+        'cast_type': cast_type,
+    }
+    if not field.company_dependent:
+        return SQL(sql_code, **sql_params)
+    # company_dependent -> jsonb
+    sql_params['field'] = SQL('value::%s', cast_type)
+    return SQL("""(
+        SELECT jsonb_object_agg(key, %(value)s)
+        FROM jsonb_each_text(%(field)s)
+    )""", field=SQL.identifier(field.name), value=SQL(sql_code, **sql_params))
```

Comment:
```suggestion
    if not field.company_dependent:
        return SQL(
            "(%(field)s - (%(factor)s - %(series_alias)s) * floor(%(total_days)s/%(factor)s) * interval '1 days')::%(cast_type)s",
            field=SQL.identifier(field.name),
            factor=factor,
            series_alias=SQL.identifier(series_alias),
            total_days=total_days,
            cast_type=cast_type,
        )
    # company_dependent -> jsonb
    return SQL(
        # jsonb_each_text returns columns 'key' and 'value'
        """ SELECT jsonb_object_agg(key,
                (value::%(cast_type)s - (%(factor)s - %(series_alias)s) * floor(%(total_days)s/%(factor)s) * interval '1 days')::%(cast_type)s)
            FROM jsonb_each_text(%(field)s) """,
        field=SQL.identifier(field.name),
        factor=factor,
        series_alias=SQL.identifier(series_alias),
        total_days=total_days,
        cast_type=cast_type,
    )
```

Reply:
You may possibly extract the formula itself as
```py
formula = "(%(value)s - (%(factor)s - %(series_alias)s) * floor(%(total_days)s/%(factor)s) * interval '1 days')::%(cast_type)s"
if not field.company_dependent:
    return SQL(
        formula,
        value=SQL.identifier(field.name),
        factor=factor,
        series_alias=SQL.identifier(series_alias),
        total_days=total_days,
        cast_type=cast_type,
    )
return SQL(
    return SQL(
        "SELECT jsonb_object_agg(key, %(expr)s) FROM jsonb_each_text(%(field)s)",
        expr=SQL(
            formula,
            value=SQL("value::%s", cast_type),
            factor=factor,
            series_alias=SQL.identifier(series_alias),
            total_days=total_days,
            cast_type=cast_type,
        ),
        field=SQL.identifier(field.name),
    )
)

Reply:
Maybe we can call the `formula` as `redistribute` 🤔

Reply:
Thanks for the suggestions, I've applied your suggestions. :)